stdlib/public/Reflection/TypeLowering.cpp

```diff
@@ -987,6 +1042,60 @@ class MultiPayloadEnumTypeInfo: public EnumTypeInfo {
   }
 };
 
+// Recursively populate the spare bit mask for this single type
+static bool populateSpareBitsMask(const TypeInfo *TI, BitMask &mask);
+
+// Recursively populate the spare bit mask for this collection of
+// record fields or enum cases.
+static bool populateSpareBitsMask(const std::vector<FieldInfo> &Fields, BitMask &mask) {
+  for (auto Field : Fields) {
+    if (Field.TR != 0) {
+      BitMask submask(Field.TI.getSize());
+      if (!populateSpareBitsMask(&Field.TI, submask)) {
+        return false;
+      }
+      mask.andMask(submask, Field.Offset);
+    }
+  }
+  return true;
+}
+
+// General recursive type walk to combine spare bit info from nested structures.
+static bool populateSpareBitsMask(const TypeInfo *TI, BitMask &mask) {
+  switch (TI->getKind()) {
+  case TypeInfoKind::Reference: {
+    // Or else somehow get access to target architecture information?
+    if (TI->getSize() == 8) {
+      // x86_64 masking (ARM64 is different)
+      mask.andMask((uint64_t)0xFF00000000000007ULL, 0);
```

Comment:
There must surely be constants for these somewhere? Or if not, we should add them?

Reply:
They're in stdlib/public/SwiftShims/System.h.  Using them is a bit tricky.  The reflection library here doesn't actually know much about the target other than the pointer size:  right now, we use the same mask for all 32-bit platforms, but not for all 64-bit, so I think the best I can do here is assume that if the pointer size is the same as the host, then it's the same architecture.  (We actually assume that a lot of places in the reflection code.)

Reply:
I've improved this:  If the target pointer size is the same as ours, assume the target is the same architecture as us.  Otherwise, assume ARM64 for 8-byte pointer target and ARM32 for 4-byte pointer targets.  The last here is fine, as all 32-bit targets currently use the same mask.  To do better for 8-byte pointer targets, we'd need some way to determine the CPU architecture of the target which I don't believe we have yet.

Reply:
We've taken a first step toward differentiating between targets in ways other than pointer size, with the ability to specify whether the target does ObjC interop. See: https://github.com/apple/swift/pull/40927/files. The right fix here would be to add another thing for x86 or ARM64 style masking. We can, of course, figure that out later.

Reply:
Would it make more sense to expose the target client triple to the host?  I guess that won't necessarily work if things like the mask style might change over time for a particular target.  Hmmm....

Reply:
Another option would be to place the mask in a variable in `Debug.h`, then read the value out of it here. As long as the runtime knows the correct mask when it's built, that would handle all scenarios.

Reply:
Hmmm....  There are four different `Debug.h` files in the source tree.  I presume you mean `include/swift/Runtime/Debug.h`

Reply:
Yeah, that's the one. We could add another `_swift_debug` variable that contains the mask.``

Reply:
The only place where I seem to be able to actually inspect the other process is from `TypeRefBuilder`.  But I don't understand the lifetime of a `TypeRefBuilder`.  It's initialized with a _reader_ object that holds a `Reader` which can actually inspect the other process.  But that `Reader` is just nonsense at initialization time -- sometimes it's NULL, sometimes it's just garbage.  So presumably `Reader` is filled in sometime later?

Reply:
The constructor for`ReflectionContext` takes a `MemoryReader` and passes it along to `super`:
```
  explicit ReflectionContext(std::shared_ptr<MemoryReader> reader)
    : super(std::move(reader), *this)
  {}
```
`super` is `MetadataReader<Runtime, TypeRefBuilder>`. Its constructor creates a `TypeRefBuilder` with the reader:
```
  template <class... T>
  MetadataReader(std::shared_ptr<MemoryReader> reader, T &&... args)
    : Builder(std::forward<T>(args)...),
      Reader(std::move(reader)),
      PtrAuthMask(queryPtrAuthMask()) {
  }
```
So everything should be set up once `TypeRefBuilder` is constructed. Not sure what would be going on there.

Reply:
The tests that are failing for me are the ones based on `swift-reflection-dump`, such as `test/Reflection/typeref_lowering.swift`

Reply:
Specifically, applying the following patch to `TypeRefBuilder.h` causes `test/Reflection/typeref_lowering.swift` to pretty reliably blow up:
```
--- a/include/swift/Reflection/TypeRefBuilder.h
+++ b/include/swift/Reflection/TypeRefBuilder.h
@@ -774,7 +774,36 @@ public:
       OpaquePointerReader([&reader](remote::RemoteAddress address, unsigned size) -> llvm::Optional<remote::RemoteAbsolutePointer> {
         return reader.Reader->readPointer(address, size);
       })
-  {}
+  {
+    if (auto Reader = reader.Reader) {
+      std::string TestName = "_swift_debug_foo";
+      auto TestAddr = Reader->getSymbolAddress(TestName);
+      uint64_t mask = 0;
+      if (TestAddr) {
+        std::cerr << "Found " << TestName << std::endl;
+      } else {
+        std::cerr << "Failed to find address of " << TestName << std::endl;
+      }
+    } else {
+      std::cerr << "No Reader?!?!" << std::endl;
+    }
+  }
 
   Demangle::Node *demangleTypeRef(RemoteRef<char> string,
                                   bool useOpaqueTypeSymbolicReferences = true) {
```

Reply:
Looks like that test uses `include/swift/StaticMirror/ObjectFileContext.h` which implements `getSymbolAddress` to always return `NULL`:
```
  // TODO: We could consult the dynamic symbol tables of the images to
  // implement this.
  reflection::RemoteAddress getSymbolAddress(const std::string &name) override {
    return reflection::RemoteAddress(nullptr);
  }
```
So that ought to work....

Reply:
I think I've addressed your concerns.  See below.